pkg/operation/botanist/secrets.go

```diff
@@ -672,22 +705,170 @@ func (b *Botanist) generateWantedSecrets(basicAuthAPIServer *secrets.BasicAuth,
+				Name: staticTokenConfig.Name,
+				Data: staticTokenSecret.Data,
+			})
+		}
+	}
+
+	gardenerResourceList = loadCertificateAuthorities(gardenerResourceList, existingSecrets, wantedCertificateAuthorityConfigs)
+	wantedShootSecrets, err := b.generateWantedSecrets(nil, nil, false, nil)
+	if err != nil {
+		return err
+	}
+
+	gardenerResourceList = loadSecrets(gardenerResourceList, existingSecrets, wantedShootSecrets)
+
+	err = kutil.TryUpdate(ctx, retry.DefaultBackoff, b.K8sGardenClient.Client(), shootState, func() error {
+		shootState.Spec.Gardener = gardenerResourceList
+		return nil
+	})
+
+	return err
+}
+
+// GenerateSecrets generates Certificate Authorities for the Shoot cluster and uses them
+// to sign certificates used by components in the Shoot's control plane. It also generates RSA key
 // pairs for SSH connections to the nodes/VMs and for the VPN tunnel. Moreover, basic authentication
 // credentials are computed which will be used to secure the Ingress resources and the kube-apiserver itself.
 // Server certificates for the exposed monitoring endpoints (via Ingress) are generated as well.
-func (b *Botanist) DeploySecrets(ctx context.Context) error {
+// All of this data is saved in the ShootState resource as it must be persisted for eventual Shoot Control Plane migration
+func (b *Botanist) GenerateSecrets(ctx context.Context) error {
+	shootState, err := b.K8sGardenClient.GardenCore().CoreV1alpha1().ShootStates(b.Shoot.Info.Namespace).Get(b.Shoot.Info.Name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	gardenerResourceList := shootState.Spec.Gardener
+	gardenerResourceDataMap, err := gardencorev1alpha1helper.CreateGardenerResourceDataMap(gardenerResourceList)
+	if err != nil {
+		return err
+	}
+
 	// If the rotate-kubeconfig operation annotation is set then we delete the existing kubecfg and basic-auth
-	// secrets. This will trigger the regeneration, incorporating new credentials. After successful deletion of all
+	// secrets and we also remove their data from gardenerResourceDataMap so that it can be regenerated and re-added to the ShootState.
+	// This will trigger the regeneration, incorporating new credentials. After successful deletion of all
 	// old secrets we remove the operation annotation.
 	if kutil.HasMetaDataAnnotation(b.Shoot.Info, common.ShootOperation, common.ShootOperationRotateKubeconfigCredentials) {
 		b.Logger.Infof("Rotating kubeconfig credentials")
-
-		for _, secretName := range []string{common.StaticTokenSecretName, common.BasicAuthSecretName, common.KubecfgSecretName} {
+		secretsToRegenerate := []string{common.StaticTokenSecretName, common.BasicAuthSecretName, common.KubecfgSecretName}
+		for _, secretName := range secretsToRegenerate {
 			if err := b.K8sSeedClient.Client().Delete(ctx, &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretName, Namespace: b.Shoot.SeedNamespace}}); client.IgnoreNotFound(err) != nil {
```

Comment:
This is because of the old implementation (which I kept) for secret object creation where we check if a secret already exists we don't try to redeploy it. I've been wondering whether I should also change that to use CreateOrUpdate(), then this extra deletion will not be necessary.